Make xDS network policies per-endpoint instead of per-identity

joestringer opened the pull request:
Since there is endpoint-specific configuration which affects policy, this PR changes network policies to be applied per endpoint (via IPs) rather than per identity.

Fixes: #3122

joestringer committed:
endpoint: Add todo to check L7 policy before update

There's a comment earlier in regeneratePolicy() that states an
expectation that no failures may occur beyond that point, then this L7
Policy push may fail after that point.

Add a comment to fix this.

Signed-off-by: Joe Stringer <joe@covalent.io>

rlenglet commented on envoy/cilium/npds.proto:
```diff
@@ -47,6 +47,10 @@ message NetworkPolicy {
+
+  // The addresses which the network policy applies to.
+  // Required.
+  repeated string addresses = 4;
```
The idea was to replace (or complement) the `uint64 policy = 1` field above with `string name = 1`, which would contain the IP address of the endpoint.

rlenglet commented on envoy/cilium/npds.proto:
```diff
@@ -71,11 +75,11 @@ message PortNetworkPolicy {
+  // in this set.
   // Optional. If not specified, any remote host is matched by this predicate.
-  repeated uint64 remote_policies = 1 [(validate.rules).repeated.unique = true];
+  repeated string remote_policies = 1 [(validate.rules).repeated.unique = true];
```
This shouldn't be changed. It's already OK.

joestringer committed:
Apply L7 network policies per endpoint

Since there is endpoint-specific configuration which affects policy,
this patch changes network policies to be applied per endpoint (via IPs)
rather than per identity.

Fixes: #3122

Signed-off-by: Joe Stringer <joe@covalent.io>

raybejjani commented on envoy/cilium/npds.proto:
```diff
@@ -32,7 +32,7 @@ service NetworkPolicyDiscoveryService {
   // The unique identifier of the network policy.
   // Required.
-  uint64 policy = 1;
+  string name = 1;
```
Is this safe? I suppose we ensure our envoy is in sync with the agent but we introduced some support for shared envoys recently. I don't expect any of our code is sensitive to field order or field number (if so, that seems ill advised). Could you make name the new #4 field? (assuming I'm correct about proto3 expectations).

joestringer replied:
For now, we ensure that the envoy and cilium versions match down to the commit SHA, so we're free to change the protobuf definition.

In future this will definitely be a concern.

rlenglet replied:
This protocol is not yet used by anything, so we can still change it in a non-backward-compatible way.

rlenglet commented on envoy/cilium/npds.proto:
```diff
@@ -47,6 +47,10 @@ message NetworkPolicy {
+
+  // The identity associated with the network policy.
+  // Optional.
+  uint64 identity = 4;
```
I would name this `policy`, to be consistent with `remote_policies`.

rlenglet commented on pkg/endpoint/endpoint.go:
```diff
@@ -1580,16 +1580,13 @@ func (e *Endpoint) LeaveLocked(owner Owner) int {
-			owner.RemoveNetworkPolicy(e.SecurityIdentity.ID)
-		}
-
+		owner.RemoveNetworkPolicy(e.IPv6, e.IPv4)
```
Do we need to keep this in the `if e.SecurityIdentity != nil {` block? Would it be safe to move it before?

joestringer replied:
I think that just means if you add an endpoint then delete it before we resolve the identity for it, then we'll make extraneous `RemoveNetworkPolicy()` calls before a NetworkPolicy is generated.

rlenglet commented on pkg/endpoint/endpoint.go:
```diff
@@ -2056,15 +2053,13 @@ func (e *Endpoint) identityLabelsChanged(owner Owner, myChangeRev int) error {
-		if lastUse {
-			owner.RemoveNetworkPolicy(e.SecurityIdentity.ID)
-		}
+		owner.RemoveNetworkPolicy(e.IPv6, e.IPv4)
```
I don't think this call makes sense anymore. We should only call this when the endpoint is destroyed, which is not the case here.

joestringer replied:
~~This is the case handling labels change on the endpoint. This could affect the policy, so I think we need to clean up the old NetworkPolicy.~~

EDIT: Thought about this again, when the next regenerate happens the new NetworkPolicy will supersede the existing one so this can be removed.

rlenglet commented on pkg/endpoint/endpoint.go:
```diff
@@ -2074,14 +2069,12 @@ func (e *Endpoint) identityLabelsChanged(owner Owner, myChangeRev int) error {
-			if lastUse {
-				owner.RemoveNetworkPolicy(e.SecurityIdentity.ID)
-			}
+			owner.RemoveNetworkPolicy(e.IPv6, e.IPv4)
```
Same as above. I don't think this call makes sense anymore. We should only call this when the endpoint is destroyed, which is not the case here.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -495,33 +495,43 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, labelsMap identity.I
-	}
+			Name:                   ipv6.String(),
+			Identity:               uint64(id),
+			IngressPerPortPolicies: getDirectionNetworkPolicy(policy.Ingress, labelsMap, deniedIngressIdentities),
```
Factorize the results of the calls to `getDirectionNetworkPolicy` to share with the IPv4 policy below.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -495,33 +495,43 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, labelsMap identity.I
-		IngressPerPortPolicies: getDirectionNetworkPolicy(policy.Ingress, labelsMap, deniedIngressIdentities),
-		EgressPerPortPolicies:  getDirectionNetworkPolicy(policy.Egress, labelsMap, deniedEgressIdentities),
-	}
+			Name:                   ipv6.String(),
```
You must check whether `ipv6` is `nil`, and return a `nil` policy in this case.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -495,33 +495,43 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, labelsMap identity.I
+			IngressPerPortPolicies: getDirectionNetworkPolicy(policy.Ingress, labelsMap, deniedIngressIdentities),
+			EgressPerPortPolicies:  getDirectionNetworkPolicy(policy.Egress, labelsMap, deniedEgressIdentities),
+		}, &cilium.NetworkPolicy{
+			Name:                   ipv4.String(),
```
You must check whether `ipv4` is `nil`, and return a `nil` policy in this case.

rlenglet commented on pkg/envoy/server.go:
```diff
@@ -495,33 +495,43 @@ func getDirectionNetworkPolicy(l4Policy policy.L4PolicyMap, labelsMap identity.I
-	name := strconv.FormatUint(uint64(id), 10)
-	NetworkPolicyCache.Delete(NetworkPolicyTypeURL, name, false)
+func RemoveNetworkPolicy(ipv6 addressing.CiliumIPv6, ipv4 addressing.CiliumIPv4) {
+	NetworkPolicyCache.Delete(NetworkPolicyTypeURL, ipv6.String(), false)
```
Check whether `ipv6` is `nil`, and don't delete in this case.
Same for `ipv4` below.
